cmd/serve.go

```diff
@@ -6,35 +6,75 @@ import (
 	"os/signal"
 	"syscall"
 
-	"github.com/odpf/salt/log"
+	"github.com/hashicorp/go-hclog"
+	hPlugin "github.com/hashicorp/go-plugin"
 	cli "github.com/spf13/cobra"
 
 	"github.com/odpf/optimus/cmd/server"
 	"github.com/odpf/optimus/config"
+	"github.com/odpf/optimus/plugin"
 )
 
-func serveCommand(l log.Logger, conf config.Optimus) *cli.Command {
-	c := &cli.Command{
+func serveCommand() *cli.Command {
+	var configFilePath string
+	cmd := &cli.Command{
 		Use:     "serve",
 		Short:   "Starts optimus service",
 		Example: "optimus serve",
 		Annotations: map[string]string{
 			"group:other": "dev",
 		},
-		RunE: func(c *cli.Command, args []string) error {
-			l.Info(coloredSuccess("Starting Optimus"), "version", config.Version)
-			optimusServer, err := server.New(l, conf)
-			defer optimusServer.Shutdown()
-			if err != nil {
-				return fmt.Errorf("unable to create server: %w", err)
-			}
-
-			sigc := make(chan os.Signal, 1)
-			signal.Notify(sigc, syscall.SIGINT, syscall.SIGTERM)
-			<-sigc
-			l.Info(coloredNotice("Shutting down server"))
-			return nil
-		},
 	}
-	return c
+
+	cmd.Flags().StringVarP(&configFilePath, "config", "c", configFilePath, "File path for client configuration")
+
+	cmd.RunE = func(c *cli.Command, args []string) error {
+		// TODO: find a way to load the config in one place
+		conf, err := config.LoadServerConfig(configFilePath)
+		if err != nil {
+			return err
+		}
+
+		// initiate jsonLogger
+		l := initLogger(jsonLoggerType, conf.Log)
+		pluginLogLevel := hclog.Info
+		if conf.Log.Level == config.LogLevelDebug {
+			pluginLogLevel = hclog.Debug
+		}
+
+		// discover and load plugins. TODO: refactor this
+		if err := plugin.Initialize(hclog.New(&hclog.LoggerOptions{
+			Name:   "optimus",
+			Output: os.Stdout,
+			Level:  pluginLogLevel,
+		})); err != nil {
+			hPlugin.CleanupClients()
+			l.Error(fmt.Sprintf("ERROR: %s\n", err.Error()))
+			os.Exit(1)
```

Comment:
Return error, running exit will not run the defer and other functions.
Please connect in case of questions.